gorm/transaction.go

```diff
@@ -57,25 +99,107 @@ func (t *Transaction) AddAfterCommitHook(hooks ...func(context.Context)) {
 	t.afterCommitHook = append(t.afterCommitHook, hooks...)
 }
 
-// BeginFromContext will extract transaction wrapper from context and start new transaction.
-// As result new instance of `*gorm.DB` will be returned.
-// Error will be returned in case either transaction or db connection info is missing in context.
-// Gorm specific error can be checked by `*gorm.DB.Error`.
-func BeginFromContext(ctx context.Context) (*gorm.DB, error) {
-	txn, ok := FromContext(ctx)
-	if !ok {
-		return nil, ErrCtxTxnMissing
+// getReadOnlyDBInstance returns the read only db txn if RO DB available otherwise it returns read/write db txn
+func getReadOnlyDBTxn(ctx context.Context, opts *databaseOptions, txn *Transaction) (*gorm.DB, error) {
+	var db *gorm.DB
+	if txn.parentRO == nil {
```

Comment:
Try to restructure the if / else / if / if / else / ... chain here as a `switch {`.

You might have to have nested levels of `switch {` or some `if` conditions inside cases, but I think it would read better e.g.

```
switch {
  case txn.parentRO == nil:
    return ...
  case opts.txOpts != nil:
    ... handle case where txn.txOpts is or isn't nil (nested if)
  case txn.txOpts != nil:
    ... at this point you know opts.txOpts is nil.
  default:
    ...
}
```